### PR TITLE
Clarify subscribe / annouce interaction

### LINKED ISSUE #362: Add text to clarify relay behavior when there are multiple announces
At interim, to support make before break use-cases, we need to clarify relay behavior when a new announce arrives for existing published.

### PR DESCRIPTION
**Pull request by fluffy**
Fixes #362 #267 #227 #225 #79

### CODE REVIEW

File: draft-ietf-moq-transport.md [hunk 1]
```diff
@@ -759,24 +768,47 @@ SUBSCRIBE_DONE, as defined below:
...
   associated with the set of tracks whose Track Namespace matches the
-  announced namespace. Specifics of where the authorization happens,
-  either at the relays or forwarded for further processing, depends on
-  the way the relay is managed and is application specific (typically
-  based on prior business agreement).
+  announced namespace. Specifics of where the authorization and
+  identification of the publisher  happens either at the relays or
+  forwarded for further processing, depends on the way the relay is
+  managed and is application specific (typically based on prior business
+  agreement).
```
Comment (afrind) [inline draft-ietf-moq-transport.md:799]: We should offer a default behavior, like 

> relays MAY have pre-negotiated mechanisms of handling authorization on behalf of publishers.  In the absence of such mechanisms, relays MUST forward SUBSCRIBE requests containing non-empty Auth Info parameters to the original publisher.

File: draft-ietf-moq-transport.md [hunk 2]
```diff
@@ -759,24 +768,47 @@ SUBSCRIBE_DONE, as defined below:
...
 Relays respond with an ANNOUNCE_OK or ANNOUNCE_ERROR control message
 providing the result of announcement. The entity receiving the
 ANNOUNCE MUST send only a single response to a given ANNOUNCE of
-either ANNOUNCE_OK or ANNOUNCE_ERROR.  When a publisher wants to stop
+either ANNOUNCE_OK or ANNOUNCE_ERROR.
+
+A Relay can receive announcements from multiple publishers for the same
+tracknamespace and it MUST respond with appropriate response to each of the
+publishers, in the same way as it would responds when processing ANNOUNCE
+from a single publisher for a given tracknamespace.
```
Comment (afrind) [inline draft-ietf-moq-transport.md:809]: Can relays limit the number of publishers for a given namespace?  It seems like for DoS reasons they can.  Maybe we should specify a minimum: "Relays MUST support at least X publishers announcing the same namespace"

File: draft-ietf-moq-transport.md [hunk 3]
```diff
@@ -788,13 +820,52 @@ match on track namespace unless otherwise negotiated by the application.
 For example, a SUBSCRIBE namespace=foobar message will be forwarded to
 the session that sent ANNOUNCE namespace=foobar.
 
+When a relay receives an incoming SUBSCRIBE for a given namespace, for
+each publisher that has announced that namespace, the relay MUST send a
+SUBSCRIBE to that publisher unless it already has an active subscription
+to that publisher for the full track name in the incoming SUBSCRIBE.
```
Comment (afrind) [inline draft-ietf-moq-transport.md:826]: What about cases where the relay can fill from cache?

File: draft-ietf-moq-transport.md [hunk 4]
```diff
@@ -710,7 +710,16 @@ subscribers for each track. Each new OBJECT belonging to the
...
+track with a SUBSCRIBE_DONE (see {{message-subscribe-done}}). A
+caching relay saves objects to its cache identified by the object's
+fulltrackname, group ID and object ID. Relays MUST be
+ready to process objects for the same fulltrackname from
+multiple publishers and the objects received are forwarded
+to active matching subscribers. If multiple objects are received
+with the same fulltrackname, group ID and object ID - a caching
+Relays MUST drop the received objects as duplicate if there exists
+already an object with the matching fulltrackame, group ID and
+object ID, in its cache.
```
Comment (afrind) [inline draft-ietf-moq-transport.md:722]: Caching and non-caching relays also need to drop (rather than forward) duplicate objects, right?

```
Pub A ---+
         +---> Relay ---> Sub
Pub B ---+
```

I think relays can figure out the best algorithms to do this, we just need to say that the subscriber only receives one copy.

Might want to add an example of why this is useful (graceful failover, redundant publishers, etc)

**Review by afrind — COMMENTED** [review]
Individual Review

This is really helpful clarification as I didn't necessarily think it had to work this way.

You have a number of random typos etc but just giving high-level feedback.

File: draft-ietf-moq-transport.md (hunk 1 above)
Comment (fluffy) [reply draft-ietf-moq-transport.md:799]: I don't think CDNs are going forward any traffic unless it is both clear who is paying for it and who is responsible for it from a DDoS point of view. We definitely need to get a better handle on that as part of the overall all issue. Glad to remove this sentence if that makes it better but just to talk about it a bit. It seems that if the CDN can look at the track name and say, yep, this is one that I bill for either by policy of by an authenticated annouce for that trackname, then it will likely forward, but if just gets a request for some random track name from some random users, seems very unlikely that it would forward that. 

At the other end of the spectrum from CDNs, a relay running in my home AP gets a a request from the local WAN side it make make complete sense for it to forward that.

File: draft-ietf-moq-transport.md (hunk 2 above)
Comment (fluffy) [reply draft-ietf-moq-transport.md:809]: well ... I suspect this is going to fall in policy for relays region. Lets say you have 10 million viewers of a Taylor Swift livestream that can all publish text comments to various chat tracks. You OK with the number of max publishers on the same namespace being in the millions ?  It seems that different relays will want very different limits here.

File: draft-ietf-moq-transport.md (hunk 3 above)
Comment (fluffy) [reply draft-ietf-moq-transport.md:826]: You mean a fetch of things that are all in the cache ? In this case definitely no subscription needed at all but I assumed that would be handled by the whatever PR eventually deals with fetch.

File: draft-ietf-moq-transport.md (hunk 4 above)
Comment (fluffy) [reply draft-ietf-moq-transport.md:722]: Lets say we had just one publisher, one high available relay, and one subscriber. I don't know how to make an HA Relay that will guarantee to never deliver the same object exactly once even during failures with any reasonable QUIC stack I have today. I think subscribers are going to have to deal with rare events. I don't think we can ask non caching relays to keep track of ever object they have forwarded to stop duplicates. 

What we certainly can do is make sure that most duplicates could be suppress particularly if the duplicated happened within the MAX_CACHE_DURATION for the object. 

I'm pretty on the fence on what to say about non caching relay. The problem really comes down to how much storage space they will use for this ( for both caching and non caching ). Clearly at some point they will free up storage and an that point any MUST is probably going to fail in some cases.   If we proposed the relay MUST cache this for at least MAX_CACHE_DURATION, I am pretty sure CDN implementers will tell us they are not going to do that 100% of the time and it should be a SHOULD. 

This is a long rambling way to get to, yah, you are right. I should have something more here. But I am thinking it will be a SHOULD level recommendation. That work for you ?

**Comment by fluffy** [comment]
> Individual Review
> 
> This is really helpful clarification as I didn't necessarily think it had to work this way.
> 
> You have a number of random typos etc but just giving high-level feedback.

Many thanks on review and any help of fixing all the small stuff (point me too it, or PR on the PR, anything) much appreciated.

File: draft-ietf-moq-transport.md (hunk 4 above)
Comment (afrind) [reply draft-ietf-moq-transport.md:722]: I think a SHOULD will probably work?  Keeping state of everything sent to every client is not scalable, but in some common cases it should be pretty easy (eg within a peep, if you see the same object ID or it goes backwards from the most recent, just drop it).  Implementation experience will help us see what's possible.

I'm thinking of a make-befrore-break case, during the point when the publisher has 2 connections to the relay, it would be nice if the relay is deduplicating as much as it can as it forwards.

File: draft-ietf-moq-transport.md (hunk 1 above)
Comment (afrind) [reply draft-ietf-moq-transport.md:799]: My concern is that without a normative default for what to do, there's very little interoperability.  The only other normative behavior I can think of is "MUST reject any subscribe with authorization info", but that seems less functional.

I 100% agree that CDNs are going to have policies and such that govern this and we should acknowledge that, but it should also work "out of the box".

We will likely revisit some of this when we tackle auth in more detail too.

File: draft-ietf-moq-transport.md (hunk 2 above)
Comment (afrind) [reply draft-ietf-moq-transport.md:809]: I was hoping SUBSCRIBE_NAMESPACE would solve some of that problem -- all 10M viewers have their own namespace `tswift/chat/unique_id`.  We may need to indicate here what limits are likely to exist, because it influences how higher level protocols will use moqt in terms of namespace and track naming.

Agreed the max will be different, but is there a min that we need to specify to get basic interop, eg: for make-before-break, min=2.

File: draft-ietf-moq-transport.md (hunk 3 above)
Comment (afrind) [reply draft-ietf-moq-transport.md:826]: I mean, 2 publishers have announced track T, and published some objects into the cache.  A subscriber issue a subscribe for track T but whatever is specified as the start and/or end is already in cache - there's no need to forward subscribe.  Maybe you can reword this to say - if you would send a subscribe to pub 1, you MUST send it to all N.

**Review by afrind — COMMENTED** [review]
Individual Comments

File: draft-ietf-moq-transport.md (hunk 4 above)
Comment (wilaw) [reply draft-ietf-moq-transport.md:722]: Rather than drop the later arriving object, I would rather that it overwrite the item existing in cache (i.e latest wins). This allows updating caches and purging corrupt content by overwriting. If we go with the proposal that the original write always wins, then we have no method for publishers to flush content from upstream caches other than waiting for its cache duration to expire.

File: draft-ietf-moq-transport.md (hunk 1 above)
Comment (wilaw) [reply draft-ietf-moq-transport.md:799]: In the absence of such mechanisms, relays MUST forward SUBSCRIBE requests containing non-empty Auth Info parameters to the original publisher.

How is that expected to work if a thousand clients make the same subscription with different auth info to an edge? Normally we would want the edge to coalesce these requests and make a single upstream subscription. But it would now have to make 1000 upstream subscriptions? I don't think that is practical. I think Cullen's proposed text is fine.

Comment (afrind) [reply draft-ietf-moq-transport.md:799]: > How is that expected to work if a thousand clients make the same subscription with different auth info to an edge?
 
 I see three possibilities:
 
 1.  pre-negotiated scheme for relay to perform auth
 2. forward the auth credentials to OP for validation 
 3. subscribe error (no way to auth)
 
If we dig into 2, right now, only SUBSCRIBE carries the auth field, so you'd end up blowing up your coalescing, unless you UNSUBSCRIBE duplicates immediately I guess.  An alternate would be to add a new message like AUTH which would perform the auth at the OP but not initiate the actual subscription.   This retains the coalescing while also providing a way for generic relays to do auth.  Maybe we can defer this to a later PR.

**Review by ianswett — COMMENTED** [review]
This looks great, and clarifies a lot about how one might use ANNOUNCE.  I have one normative comment, which is that I think some of the MUSTs should be a SHOULD.

**Review by suhasHere — APPROVED** [review]
This clarifies lot of ambiguities on implementing make before break use-cases.
LGTM

File: draft-ietf-moq-transport.md [hunk 5]
```diff
@@ -710,7 +710,16 @@ subscribers for each track. Each new OBJECT belonging to the
 track within the subscription range is forwarded to each active
 subscriber, dependent on the congestion response. A subscription
 remains active until the publisher of the track terminates the
-track with a SUBSCRIBE_DONE (see {{message-subscribe-done}}).
+subscription with a SUBSCRIBE_DONE (see {{message-subscribe-done}}). A
```
Comment (martinduke) [inline draft-ietf-moq-transport.md:713]: or UNSUBSCRIBE? Or is there a SUBSCRIBE_DONE in reply?

File: draft-ietf-moq-transport.md [hunk 6]
```diff
@@ -759,24 +768,46 @@ SUBSCRIBE_DONE, as defined below:
...
+all active subscriptions to the new relay. The new relay will send a
+response to the subscribes and once this has happened, the subscription
+to the old relay can be stopped with an UNSUBSCRIBE.
+
+
 ## Publisher Interactions
 
 Publishing through the relay starts with publisher sending ANNOUNCE
 control message with a `Track Namespace` ({{model-track}}).
+The annouce allows the relays to know which publisher to forward a
```
Comment (martinduke) [inline draft-ietf-moq-transport.md:785]:
```suggestion
The announce allows the relays to know which publisher to forward a
```

File: draft-ietf-moq-transport.md [hunk 7]
```diff
@@ -788,13 +819,51 @@ match on track namespace unless otherwise negotiated by the application.
 For example, a SUBSCRIBE namespace=foobar message will be forwarded to
 the session that sent ANNOUNCE namespace=foobar.
 
+When a relay receives an incoming SUBSCRIBE for a given namespace, for
+each publisher that has announced that namespace, the relay MUST send a
+SUBSCRIBE to that publisher unless it already has an active subscription
```
Comment (martinduke) [inline draft-ietf-moq-transport.md:822]: Unless the entire range is already cached?

File: draft-ietf-moq-transport.md [hunk 8]
```diff
@@ -788,13 +819,51 @@ match on track namespace unless otherwise negotiated by the application.
...
+to that publisher for the full track name in the incoming SUBSCRIBE.
+
+When a relay receives an incoming ANNOUCE for a given namespace, for
+each active subscription that matches that namespace, it MUST send a
+SUBSCRIBE to that publisher that send the ANNOUCE.
```
Comment (martinduke) [inline draft-ietf-moq-transport.md:827]:
```suggestion
SUBSCRIBE to that publisher that send the ANNOUNCE.
```

File: draft-ietf-moq-transport.md [hunk 9]
```diff
@@ -788,13 +820,52 @@ match on track namespace unless otherwise negotiated by the application.
 For example, a SUBSCRIBE namespace=foobar message will be forwarded to
 the session that sent ANNOUNCE namespace=foobar.
 
+When a relay receives an incoming SUBSCRIBE for a given namespace, for
+each publisher that has announced that namespace, the relay MUST send a
```
[the remark this answers is not in the document]
Comment (martinduke) [reply draft-ietf-moq-transport.md:821]: +1

**Review by martinduke — CHANGES_REQUESTED** [review]

File: draft-ietf-moq-transport.md (hunk 1 above)
Comment (fluffy) [reply draft-ietf-moq-transport.md:796]: Going to leave the auth stuff for a later work on auth for the whole draft

File: draft-ietf-moq-transport.md (hunk 7 above)
Comment (fluffy) [reply draft-ietf-moq-transport.md:825]: yah .... let me see what to put here. A bunch of this changes depending on how subscribe is also fetch or if they are separate.

Comment (fluffy) [reply draft-ietf-moq-transport.md:826]: I think the text now deals with that

File: draft-ietf-moq-transport.md (hunk 5 above)
Comment (fluffy) [reply draft-ietf-moq-transport.md]: I'm adding "or all the downstream subscriptions are ended". I started with just SUBSCRIBE_DONE  but there are also all the timeout and connection close things that can end a subscription. The draft overall is a bit light on all the clean up and timeouts of things and I can't fix that in this PR.

File: draft-ietf-moq-transport.md [hunk 10]
```diff
@@ -759,24 +768,47 @@ SUBSCRIBE_DONE, as defined below:
...
+
+A Relay can receive announcements from multiple publishers for the same
+tracknamespace and it MUST respond with appropriate response to each of the
+publishers, in the same way as it would responds when processing ANNOUNCE
+from a single publisher for a given tracknamespace.
+
+When a publisher wants to stop
 new subscriptions for an announced namespace it sends an UNANNOUNCE.
 A subscriber indicates it will no longer route subscriptions for a
 namespace it previously responded ANNOUNCE_OK to by sending an
```
[the remark this answers is not in the document]
Comment (fluffy) [reply draft-ietf-moq-transport.md:815]: I've got no idea :-) We need an error to ANNOUNCE for auth errors, so it makes sense to have an OK just to know there was no error. But I have no idea why we need CANCEL

File: draft-ietf-moq-transport.md (hunk 3 above)
Comment (fluffy) [reply draft-ietf-moq-transport.md]: I like that - added in as you sugesteed.

**Comment by fluffy** [comment]
@martinduke - thanks for review. I think I fixed all the stuff you raised. Can you take another look 

@ianswett - handing the baton to you on this one. Thanks you

File: draft-ietf-moq-transport.md [hunk 11]
```diff
@@ -710,7 +710,18 @@ subscribers for each track. Each new OBJECT belonging to the
 track within the subscription range is forwarded to each active
 subscriber, dependent on the congestion response. A subscription
 remains active until the publisher of the track terminates the
-track with a SUBSCRIBE_DONE (see {{message-subscribe-done}}).
+subscription with a SUBSCRIBE_DONE (see {{message-subscribe-done}})
+or all the downstream subscriptions are ended. A
```
Comment (martinduke) [inline draft-ietf-moq-transport.md:714]: The "publisher" is a term local to the session, unless you mean "original publisher." So a subscription ends when the publisher sends SUBSCRIBE_DONE or when the subscriber sends UNSUBSCRIBE. Downstream subscriptions don't enter into it.

File: draft-ietf-moq-transport.md [hunk 12]
```diff
@@ -759,24 +770,46 @@ SUBSCRIBE_DONE, as defined below:
 | 0x6  | Expired                   |
 |------|---------------------------|
 
+### Graceful Publisher Relay Switchover
+
+This section describes behavior a subscriber MAY implement
+to allow for a better user experience when a relay sends a GOAWAY.
+
+When a subscriber receives the GOAWAY message, it starts of the process
```
Comment (martinduke) [inline draft-ietf-moq-transport.md:778]:
```suggestion
When a subscriber receives the GOAWAY message, it starts the process
```

File: draft-ietf-moq-transport.md [hunk 13]
```diff
@@ -759,24 +770,46 @@ SUBSCRIBE_DONE, as defined below:
...
 
 Relays respond with an ANNOUNCE_OK or ANNOUNCE_ERROR control message
 providing the result of announcement. The entity receiving the
 ANNOUNCE MUST send only a single response to a given ANNOUNCE of
-either ANNOUNCE_OK or ANNOUNCE_ERROR.  When a publisher wants to stop
+either ANNOUNCE_OK or ANNOUNCE_ERROR.
+
+A Relay can receive announcements from multiple publishers for the same
+tracknamespace and it MUST respond with appropriate response to each of the
+publishers, in the same way as it would responds when processing ANNOUNCE
```
Comment (martinduke) [inline draft-ietf-moq-transport.md:809]:
```suggestion
publishers, in the same way as it would respond when processing ANNOUNCE
```

Comment (martinduke) [reply draft-ietf-moq-transport.md:809]: alternately, "it responds"

File: draft-ietf-moq-transport.md [hunk 14]
```diff
@@ -788,13 +821,52 @@ match on track namespace unless otherwise negotiated by the application.
 For example, a SUBSCRIBE namespace=foobar message will be forwarded to
 the session that sent ANNOUNCE namespace=foobar.
 
+When a relay receives an incoming SUBSCRIBE for a given namespace
+that would cause it to send an upstream subscription, for
+each publisher that has announced that namespace, the relay SHOULD send a
+SUBSCRIBE to that publisher unless it already has an active subscription
+to that publisher for the Full Track Name in the incoming SUBSCRIBE.
```
Comment (martinduke) [inline draft-ietf-moq-transport.md:828]: s/for the Full Track Name/for the objects in the specified range

File: draft-ietf-moq-transport.md [hunk 15]
```diff
@@ -788,13 +821,52 @@ match on track namespace unless otherwise negotiated by the application.
...
+
+When a relay receives an incoming ANNOUCE for a given namespace, for
+each active subscription that matches that namespace, it SHOULD send a
+SUBSCRIBE to that publisher that send the ANNOUNCE.
```
Comment (martinduke) [inline draft-ietf-moq-transport.md:832]: I think this means that if there is an ANNOUNCE for a track where a relay has N incoming subscriptions, it should send N SUBSCRIBEs. That seems to negate the whole purpose of relays?

I hope you actually mean the relay should send *one* SUBSCRIBE assuming the ranges are the same. If not, we are not on the same page at all.

File: draft-ietf-moq-transport.md [hunk 16]
```diff
@@ -788,13 +821,52 @@ match on track namespace unless otherwise negotiated by the application.
 For example, a SUBSCRIBE namespace=foobar message will be forwarded to
 the session that sent ANNOUNCE namespace=foobar.
 
+When a relay receives an incoming SUBSCRIBE for a given namespace
+that would cause it to send an upstream subscription, for
+each publisher that has announced that namespace, the relay SHOULD send a
+SUBSCRIBE to that publisher unless it already has an active subscription
```
Comment (afrind) [inline draft-ietf-moq-transport.md:827]: meta.ai suggests:

When a relay receives an incoming SUBSCRIBE request for a given namespace that would trigger an upstream subscription, it SHOULD send a SUBSCRIBE request to each publisher that has announced that namespace, unless it already has an active subscription to that publisher for the Full Track Name specified in the incoming SUBSCRIBE request.

File: draft-ietf-moq-transport.md (hunk 15 above)
Comment (afrind) [reply draft-ietf-moq-transport.md:832]: Maybe "each active *upstream* subscription" would help clarify?  

I think some of this language will need a second pass once we sort fetch and ranges.

File: draft-ietf-moq-transport.md [hunk 17]
```diff
@@ -788,13 +821,52 @@ match on track namespace unless otherwise negotiated by the application.
...
+to that publisher for the Full Track Name in the incoming SUBSCRIBE.
+
+When a relay receives an incoming ANNOUCE for a given namespace, for
+each active subscription that matches that namespace, it SHOULD send a
```
Comment (ianswett) [inline draft-ietf-moq-transport.md:831]:
```suggestion
each active upstream subscription that matches that namespace, it SHOULD send a
```

File: draft-ietf-moq-transport.md (hunk 15 above)
Comment (ianswett) [reply draft-ietf-moq-transport.md:832]: I added upstream.

**Review by ianswett — APPROVED** [review]

File: draft-ietf-moq-transport.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (suhasHere) [reply draft-ietf-moq-transport.md:821]: ANNOUNCE_CANCEL  is used in the below flow IIYC

```mermaid
sequenceDiagram
    Publisher->>Relay(Subscriber): Announce
    Relay(Subscriber)->>Publisher: Announce Ok
    Relay(Subscriber)->>Publisher: Subscribe
    Publisher->>Relay(Subscriber): Subscribe OK
    Publisher->Relay(Subscriber): Objects Flow
    Note right of Relay(Subscriber):  Relays is unable to support new subscriptions 
    Relay(Subscriber)->>Publisher: Announce Cancel 
```

Existing subscriptions will still get served, but new subscriptions will not be sent to the publisher anymore.

File: draft-ietf-moq-transport.md (hunk 15 above)
Comment (fluffy) [reply draft-ietf-moq-transport.md:838]: Thanks - yes, upstream
